keanu-project/src/test/java/io/improbable/keanu/algorithms/variational/KLDivergenceTest.java

```diff
@@ -0,0 +1,112 @@
+
+import io.improbable.keanu.algorithms.NetworkSamples;
+import io.improbable.keanu.algorithms.mcmc.MetropolisHastings;
+import io.improbable.keanu.network.BayesianNetwork;
+import io.improbable.keanu.network.NetworkState;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.VertexId;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.ConstantDoubleVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.GaussianVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.ProbabilisticDouble;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class KLDivergenceTest {
+
+    @Test
+    public void klDivergenceIsZeroIfPAndQAreIdentical() {
+        double identicalProb = Math.log(0.5);
+
+        TestQDistribution q = new TestQDistribution(identicalProb);
+        NetworkSamples samples = createNetworkSamplesWithOneVertexAndOneSample(identicalProb);
+
+        assertEquals(0., KLDivergence.compute(q, samples), 1e-6);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void throwsExceptionIfPProbIsNotZeroButQIs() {
+        TestQDistribution q = new TestQDistribution(Double.NEGATIVE_INFINITY);
+        NetworkSamples samples = createNetworkSamplesWithOneVertexAndOneSample(Math.log(0.5));
+
+        KLDivergence.compute(q, samples);
+    }
+
+    @Test
+    public void klDivergenceReturnsZeroIfPIsZeroButQIsNot() {
+        TestQDistribution q = new TestQDistribution(Math.log(0.5));
+        NetworkSamples samples = createNetworkSamplesWithOneVertexAndOneSample(Double.NEGATIVE_INFINITY);
+
+        assertEquals(0., KLDivergence.compute(q, samples), 1e-10);
+    }
+
+    @Test
+    public void klDivergenceReturnsZeroIfPAndQAreZero() {
+        TestQDistribution q = new TestQDistribution(Double.NEGATIVE_INFINITY);
+        NetworkSamples samples = createNetworkSamplesWithOneVertexAndOneSample(Double.NEGATIVE_INFINITY);
+
+        assertEquals(0., KLDivergence.compute(q, samples), 1e-10);
+    }
+
+    @Test
+    public void returnsLargerKLDivergenceIfTheLocationOfQIsFurtherFromP() {
+        GaussianVertex v1 = new GaussianVertex(0., 1.);
+        ConstantDoubleVertex v2 = new ConstantDoubleVertex(0.1);
+        DoubleVertex v3 = v1.plus(v2);
+
+        BayesianNetwork network = new BayesianNetwork(v3.getConnectedGraph());
+        NetworkSamples samples = MetropolisHastings
+            .withDefaultConfig()
+            .getPosteriorSamples(network, Collections.singletonList(v1), 1000);
+
+        ProbabilisticDouble q1 = new GaussianVertex(0.1, 1.);
+        ProbabilisticDouble q2 = new GaussianVertex(10.0, 1.);
+
+        assertTrue(KLDivergence.compute(q1, samples) < KLDivergence.compute(q2, samples));
```

Comment:
It's a good idea to use Hamcrest matchers for asserts. The code reads more fluently and the error messages are much more informative.

For example:

    assertThat(KLDivergence.compute(q1, samples), lessThan(KLDivergence.compute(q2, samples)));

yields messages like:

    java.lang.AssertionError: 
    Expected: a value less than <14325.522530625412>
         but: <1435844.110279129> was greater than <14325.522530625412>

Whereas the `assertTrue` statement only tells you that there's an `java.lang.AssertionError`, with no message.